shared/lib-hoppy/hoppy/async_consumer.py

```diff
@@ -0,0 +1,137 @@
+
+        logging.debug(f'event=specifyingQualityOfService '
+                      f'client_type={self._client_type} '
+                      f'prefetch_count={self._prefetch_count}')
+        self._channel.basic_qos(prefetch_count=self._prefetch_count, callback=self._on_basic_qos_ok)
+
+    def _shut_down(self):
+        """Called when the client is requested to stop.
+        Overrides super class abstract method"""
+
+        if self._consuming:
+            self._stop_consuming()
+
+    def _on_basic_qos_ok(self, _unused_frame):
+        logging.debug(f'event=specifiedQualityOfService '
+                      f'client_type={self._client_type} '
+                      f'prefetch_count={self._prefetch_count}')
+        self._start_consuming()
+
+    def _start_consuming(self):
+        self._channel.add_on_cancel_callback(self._on_consumer_cancelled)
+        self._consumer_tag = self._channel.basic_consume(self.queue_name, self._on_message)
+        self._consuming = True
+        logging.debug(f'event=startConsuming '
+                      f'client_type={self._client_type} '
+                      f'consumer_tag={self._consumer_tag}')
+
+    def _on_consumer_cancelled(self, method_frame):
+        logging.debug(f'event=serverCancelledConsumer '
+                      f'client_type={self._client_type} '
+                      f'method_frame={method_frame}')
+        super()._close_channel()
+
+    def _stop_consuming(self):
+        if self._channel:
+            logging.debug(f'event=stoppingConsuming '
+                          f'client_type={self._client_type} '
+                          f'consumer_tag={self._consumer_tag}')
+            self._channel.basic_cancel(self._consumer_tag, self._on_cancel_ok)
+
+    def _on_cancel_ok(self, _unused_frame):
+        self._consuming = False
+        logging.debug(f'event=stoppedConsuming '
+                      f'client_type={self._client_type} '
+                      f'consumer_tag={self._consumer_tag}')
+        self._close_channel()
+
+    def _on_message(self, _unused_channel, basic_deliver, properties, body):
+        logging.debug(f'event=receivedMessage '
+                      f'client_type={self._client_type} '
+                      f'app_id={properties.app_id} '
+                      f'delivery_tag={basic_deliver.delivery_tag} '
+                      f'correlation_id={properties.correlation_id}')
+        if self.reply_callback is not None:
+            try:
+                self.reply_callback(self._channel,
+                                    properties,
+                                    basic_deliver.delivery_tag,
+                                    body)
+            except Exception as e:
+                logging.error(f'event=couldNotCallConsumeCallback '
+                              f'client_type={self._client_type} '
+                              f'callback={self.reply_callback} '
+                              f'err={e}')
```

Comment:
When converting Python exceptions to strings, you generally want to use `repr(e)` or `{e!r}`, instead of `str(e)` or `{e}`, because the repr is a more faithful representation of the argument. For example,

```python
try: {}[123]
except Exception as e: print(str(e), repr(e))
```

will print `123` as the str and `KeyError(123)` as the repr.